File: Python/QiPy/QiClient.py
from urlparse import urlparse
import urllib
import httplib as http
import json
from QiType import QiType
from QiStream import QiStream
from QiStreamBehavior import QiStreamBehavior
from QiError import QiError
from JsonEncoder import Encoder
import requests
import time

class QiClient(object):
    """description of class"""

    def __init__(self, url, authItems):
        self.url = url
        version = 0.1
        self.__tenantsBase = "/Qi/Tenants"
        self.__typesBase = "/Qi/Types"
        self.__streamsBase = "/Qi/Streams"
        self.__behaviorBase = "/Qi/Behaviors"
        self.__insertSingle = "/Data/InsertValue"
        self.__insertMultiple = "/Data/InsertValues"
        self.__getTemplate = "/{stream_id}/Data/GetWindowValues?{start}&{end}"
        self.__getRangeTemplate = "/Data/GetRangeValues?startIndex={start}&skip={skip}&count={count}&reversed={reverse}&boundaryType={boundaryType}"
        self.__updateSingle = "/Data/UpdateValue"
        self.__updateMultiple = "/Data/UpdateValues"
        self.__removeSingleTemplate = "/{stream_id}/Data/RemoveValue?{param}"
        self.__removeMultipleTemplate = "/{stream_id}/Data/RemoveWindowValues?{start}&{end}"
        self.__getLast = "/data/getlastvalue"
        self.__replaceValue = "/data/replaceValue"
        self.__replaceValues = "/data/replaceValues"
        self.__authItems = authItems
        self.__token = ""
        self.__expiration = ""
        
        self.__getToken()
        if not self.__token:
            return

        print "----------------------------------"
        print "  ___  _ ____"        
        print " / _ \(_)  _ \ _   _ "
        print "| | | | | |_) | | | |"
        print "| |_| | |  __/| |_| |"
        print " \__\_\_|_|    \__, |"
        print "               |___/ "	
        print "Version "+ str(version)
        print "----------------------------------"
        #conn = http.HTTPConnection(self.url)
        #conn.request("HEAD", "/Qi/Types", headers = self.__qi_headers())
        #response = conn.getresponse()        
        #if response.status != 200:
        #    raise QiError("Failed to connect to Qi endpoint {status}:{reason}".
        #                  format(status = response.status, reason = response.reason))

        print "Qi endpoint at {url}".format(url = self.url)

    
    # Qi Type
    
    def createType(self, qi_type):
        if qi_type is None:
            return
       
        if not isinstance(qi_type, QiType):
            return
        
        self.__getToken()
        
        conn = http.HTTPSConnection(self.url)
        
        conn.request("POST", self.__typesBase, qi_type.toString(), self.__qi_headers())
        response = conn.getresponse()
        
        if response.status == 302: # Found                      
            url = urlparse(response.getheader("Location"))
            conn = http.HTTPSConnection(self.url)
            conn.request("GET", url.path, headers = self.__qi_headers())            
            response = conn.getresponse()
        
        if response.status == 200 or response.status == 201:
            type = QiType.fromString(response.read().decode())
            conn.close()
            return type
        else:
            conn.close()
            raise QiError("Failed to create type. {status}:{reason}".format(status = response.status, reason = response.reason))
    
    def getTypes(self):
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        
        conn.request("GET", self.__typesBase, headers = self.__qi_headers())
        response = conn.getresponse()
        
        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get QiTypes {status}:{reason}".
                          format(status = response.status, reason = response.reason))
        typesresponse = response.read().decode()
        conn.close()
        typeslist = json.loads(typesresponse)
        returnlist = []
        for typedict in typeslist:
            returnlist.append(QiType.fromDictionary(typedict))
        
        return returnlist

    def getType(self, type_id):
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        
        conn.request("GET", self.__typesBase + '/' + type_id, headers = self.__qi_headers())
        response = conn.getresponse()
        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get QiType, {type_id}. {status}:{reason}".
                          format(type_id = type_id, status = response.status, reason = response.reason))
        
        typesresponse = response.read().decode()
        conn.close()
        return QiType.fromString(typesresponse)        
                
    def deleteType(self, type_id):
        self.__getToken()
        if type_id is None:
            return
        conn = http.HTTPSConnection(self.url)
        conn.request('DELETE', self.__typesBase + '/' +  type_id, headers = self.__qi_headers())
        response = conn.getresponse()
        
        if response.status != 200:
            conn.close()
            raise QiError("Failed to delete QiType, {type_id}. {status}:{reason}".
                          format(type_id = type_id, status = response.status, reason = response.reason))

    def createBehavior(self, behavior):
        self.__getToken()
        if not isinstance(behavior, QiStreamBehavior):
            return
        conn = http.HTTPSConnection(self.url)
        
        payload = behavior.toString()
        conn.request("POST", self.__behaviorBase, payload, self.__qi_headers())
        response = conn.getresponse()

        if response.status == 302: # Found                      
            url = urlparse(response.getheader("Location"))
            conn = http.HTTPSConnection(self.url)
            conn.request("GET", url.path, headers = self.__qi_headers())            
            response = conn.getresponse()
        if response.status == 200 or response.status == 201:
            string = response.read().decode()
            type = QiStreamBehavior.fromString(string)
            conn.close()
            return type
        else:
            conn.close()
            raise QiError("Failed to create behavior. {status}:{reason}".format(status = response.status, reason = response.reason))
            
    def deleteBehavior(self, behaviorId):
        self.__getToken()
        if behaviorId is None:
            return
        conn = http.HTTPSConnection(self.url)
        conn.request('DELETE', self.__behaviorBase + '/' +  behaviorId, headers = self.__qi_headers())
        response = conn.getresponse()
        
        if response.status != 200:
            conn.close()
            raise QiError("Failed to delete QiType, {behaviorId}. {status}:{reason}".
                            format(behaviorId = behaviorId, status = response.status, reason = response.reason))
                          
    # Qi Stream
    def createStream(self, qi_stream):
        self.__getToken()
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            return
   
        conn = http.HTTPSConnection(self.url)
        
        conn.request("POST", self.__streamsBase, qi_stream.toString(), self.__qi_headers())

        response = conn.getresponse()
        
        if response.status == 302:
            url =  urlparse(response.getheader("Location"))
            conn = http.HTTPSConnection(self.url)
            conn.request("GET", url.path, headers = self.__qi_headers())
            response = conn.getresponse()
        if response.status == 200 or response.status == 201:
            stream = QiStream.fromString(response.read().decode())
            conn.close()
            return stream
        else:
            conn.close()
            raise QiError("Failed to create QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
        

    def getStream(self, stream_id):
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        
        conn.request("GET", self.__streamsBase + '/' + stream_id, headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get QiStream {stream_id}. {status}:{reason}".
                          format(stream_id = stream_id, status = response.status, reason = response.reason))
        
        streamResponse = response.read().decode()
        conn.close()
        return QiStream.fromDictionary(json.loads(streamResponse))        

    def getStreams(self):
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        
        conn.request("GET", self.__streamsBase, headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get QiStreams. {status}:{reason}".
                          format(status = response.status, reason = response.reason))
        
        streamssresponse = response.read().decode()
        conn.close()
        streamsslist = json.loads(streamssresponse)
        returnlist = []
        for streamdict in streamsslist:
            returnlist.append(QiStream.fromDictionary(streamdict))
        
        return returnlist

    def updateStream(self, qi_stream): 
        self.__getToken()
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            return
        
        conn = http.HTTPSConnection(self.url)
        conn.request("PUT", self.__streamsBase + '/' + qi_stream.Id, qi_stream.toString(), self.__qi_headers())

        response = conn.getresponse()
     
        if response.status != 200:
            conn.close()
            raise QiError("Failed to edit QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
        conn.close()


    def deleteStream(self, stream_id):
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        
        conn.request("DELETE", self.__streamsBase + '/' + stream_id, headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to delete QiStream {stream_id}. {status}:{reason}".
                          format(stream_id = stream_id, status = response.status, reason = response.reason))
        
        conn.close()
    

    # Stream data

    def getLastValue(self, qi_stream):
        if qi_stream is None:
            return
        
        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        conn = http.HTTPSConnection(self.url)
        conn.request("GET", self.__streamsBase + '/' + qi_stream.Id + self.__getLast, 
                     headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get last value for QiStream {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
        
        streamResponse = response.read().decode()
        conn.close()
        return json.loads(streamResponse)

    def insertValue(self, qi_stream, value):
        """
        Insert the specified value into this Qi stream.
        The value parameter must be a dictionary of data matching
        the stream type. For example:
        {"Value": 102.3, "Timestamp": "2015-08-12T22:07:49.4472922Z" }
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(value, cls = Encoder)
        
        conn = http.HTTPSConnection(self.url)
        conn.request("POST", self.__streamsBase + '/' + qi_stream.Id + self.__insertSingle, 
                     payload, self.__qi_headers())

        response = conn.getresponse()
        
        if response.status != 200:
            conn.close()
            raise QiError("Failed to insert value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
                
    def updateValue(self, qi_stream, value):
        """
        Update a single event if found
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(value, cls = Encoder)
        
        conn = http.HTTPSConnection(self.url)
        conn.request("PUT", self.__streamsBase + '/' + qi_stream.Id + self.__updateSingle, 
                     payload, self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to replace value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))

    def updateValues(self, qi_stream, value):
        """
        Update an arry of events in a Qi stream. Replaces an event with the same key.
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(value, cls = Encoder)

        conn = http.HTTPSConnection(self.url)
        conn.request("PUT", self.__streamsBase + '/' + qi_stream.Id + self.__updateMultiple, 
                     payload, self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to replace value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
                          
    def replaceValue(self, qi_stream, value):
        """
        Update the specified value, as matched by *key* type property
        into this Qi stream. The value parameter must be a dictionary of data matching
        the stream type. For example:
        {"Value": 102.3, "Timestamp": "2015-08-12T22:07:49.4472922Z" }
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(value)

        conn = http.HTTPSConnection(self.url)
        conn.request("PUT", self.__streamsBase + '/' + qi_stream.Id + self.__replaceValue, 
                     payload, self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to replace value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
            
    def removeValue(self, qi_stream, index):
        """
        Delete the value at the specified index. The value index is the data stored
        in the *key* type property of the Qi type. The index parameter must be a 
        string. For example:
        '2015-08-12T22:07:49.4472922Z'
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        params = urllib.urlencode({"index": index})
        conn = http.HTTPSConnection(self.url)
        conn.request("DELETE", self.__streamsBase + '/' + self.__removeSingleTemplate.format(stream_id = qi_stream.Id, param = params), 
                     headers = self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to remove value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
          
        
    def insertValues(self, qi_stream, values):
        """
        Insert the specified list of values submitted into the specified Qi stream.
        The values parameter must be a list of dictionaries like:
        [{"Value": 102.3, "Timestamp": "2015-08-12T22:07:49.111000Z" }, 
        {"Value": 104.5, "Timestamp": "2015-08-12T24:07:49.9991111Z" }]
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(values, cls = Encoder)
        
        conn = http.HTTPSConnection(self.url)
        conn.request("POST", self.__streamsBase + '/' + qi_stream.Id + self.__insertMultiple, 
                     payload, self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to insert values for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
    
    def replaceValues(self, qi_stream, value):
        """
        Update the specified values, as matched by *key* type property
        into this Qi stream. The value must be an array of events whose keys should be updated
        [{"Value": 102.3, "Timestamp": "2015-08-12T22:07:49.4472922Z" },
        {"Value": 101.1, "Timestamp": "2015-08-12T22:08:49.4472922Z" }]
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        payload = json.dumps(value)
        conn = http.HTTPSConnection(self.url)
        conn.request("PUT", self.__streamsBase + '/' + qi_stream.Id + self.__replaceValues, 
                     payload, self.__qi_headers())

        response = conn.getresponse()

        if response.status != 200:
            conn.close()
            raise QiError("Failed to replace value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
    
    def removeValues(self, qi_stream, start, end):
        """
        Delete the value at the specified index. The value index is the data stored
        in the *key* type property of the Qi type. The index parameter must be a 
        string. For example:
        '2015-08-12T22:07:49.4472922Z'
        """
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        #params = urllib.urlencode({"startIndex": start, "endIndex": end}) 
        conn = http.HTTPSConnection(self.url)
        conn.request("DELETE", self.__streamsBase + '/' + 
                        self.__removeMultipleTemplate.format(stream_id = qi_stream.Id, 
                        start = urllib.urlencode({"startIndex": start}),
                        end = urllib.urlencode({"endIndex": end})), 
                        headers = self.__qi_headers())

        response = conn.getresponse()
        if response.status != 200:
            conn.close()
            raise QiError("Failed to remove value for QiStream, {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
                          
    def getWindowValues(self, qi_stream, start, end):
        if qi_stream is None:
            return

        if not isinstance(qi_stream, QiStream):
            raise TypeError("stream must be a valid QiStream")
            
        self.__getToken()
        conn = http.HTTPSConnection(self.url)

        conn.request("GET", self.__streamsBase + '/' + 
                        self.__getTemplate.format(stream_id = qi_stream.Id, 
                                                 start = urllib.urlencode({"startIndex": start}), 
                                                    end = urllib.urlencode({"endIndex": end})), 
                        headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get last value for QiStream {stream_id}. {status}:{reason}".
                          format(stream_id = qi_stream.Id, status = response.status, reason = response.reason))
        
        streamResponse = response.read().decode()
        conn.close()
        return json.loads(streamResponse)

    def getRangeValues(self, streamId, start, skip, count, reverse, boundaryType):
            
        self.__getToken()
        conn = http.HTTPSConnection(self.url)

        conn.request("GET", self.__streamsBase + '/' + 
                        streamId +
                        self.__getRangeTemplate.format(start = start, 
                                                    skip = str(skip),
                                                    count = str(count),
                                                    reverse = str(reverse),
                                                    boundaryType = str(boundaryType)), 
                        headers = self.__qi_headers())
        response = conn.getresponse()

        if response.status != 200:            
            conn.close()
            raise QiError("Failed to get last value for QiStream {stream_id}. {status}:{reason}".
                          format(stream_id = streamId, status = response.status, reason = response.reason))
        
        streamResponse = response.read().decode()
        conn.close()
        return json.loads(streamResponse)

    #get all Qi types
    def listTypes(self):
        types = self.getTypes()
        print "{len} Qi types found:".format(len = len(types))
        print ",\n ".join(t.Name for t in types)
    
    #list all Qi streams
    def listStreams(self):
        streams = self.getStreams()
        print "{len} Qi streams found:".format(len = len(streams))
        print ",\n".join("{0} ({1})".format(t.Name, t.TypeId)  for t in streams)
        
    # private methods
    def __getToken(self):     
        if (self.__expiration and self.__expiration < (time.time()/100)):
            return
        response = requests.post(self.__authItems['authority'], 
                                 data = { 'grant_type' : 'client_credentials',
                                         'client_id' : self.__authItems['appId'],
                                            'client_secret' : self.__authItems['appKey'],
                                            'resource' : self.__authItems['resource']
                                            })
        if response.status_code == 200:
            self.__token = response.json()['access_token']
            self.__expiration = response.json()['expires_on']
        else:
            self.__token = ""
            print "Authentication Failure : "+response.reason
            
    def __qi_headers(self):
        return {
            "Authorization" : "bearer %s" % self.__token,
            "Content-type": "application/json", 
            "Accept": "text/plain"
            }

File: JavaScript/QiClient.js


//function to create client object
var qiObjs = require("./QiObjects.js");
var restCall = require("request-promise");

String.prototype.format = function (args) {
            var str = this;
            return str.replace(String.prototype.format.regex, function(item) {
                var intVal = parseInt(item.substring(1, item.length - 1));
                var replace;
                if (intVal >= 0) {
                    replace = args[intVal];
                } else if (intVal === -1) {
                    replace = "{";
                } else if (intVal === -2) {
                    replace = "}";
                } else {
                    replace = "";
                }
                return replace;
            });
        };
        String.prototype.format.regex = new RegExp("{-?[0-9]+}", "g");

module.exports = {
    QiClient : function(url){
    	this.url = url;
        this.version = 0.1;
        this.tenantsBase = "/Qi/Tenants";
        this.typesBase = "/Qi/Types";
        this.streamsBase = "/Qi/Streams";
        this.behaviorBase = "/Qi/Behaviors";
        this.insertSingle = "/Data/InsertValue";
        this.insertMultiple = "/Data/InsertValues";
        this.getTemplate = "/{0}/Data/GetWindowValues?startIndex={1}&endIndex={2}";
        this.getRangeTemplate = "/{0}/Data/GetRangeValues?startIndex={1}&skip={2}&count={3}&reversed={4}&boundaryType={5}";
        this.updateSingle = "/Data/UpdateValue";
        this.updateMultiple = "/Data/UpdateValues";
        this.removeSingleTemplate = "/{0}/Data/RemoveValue?index={1}";
        this.removeMultipleTemplate = "/{0}/Data/RemoveWindowValues?startIndex={1}&endIndex={2}";
        this.getLast = "/data/getlastvalue";
        this.replaceValue = "/data/replaceValue";
        this.replaceValues = "/data/replaceValues";
        this.tenantId = "localtenant";
        this.token = "";
        this.tokenExpires = "";

        this.getToken = function(authItems){
                                return restCall({
                                                url : authItems["authority"],
                                                method: 'POST',
                                                headers : {
                                                            'Content-Type': 'application/x-www-form-urlencoded'
                                                },
                                                form : {    
                                                            'grant_type' : 'client_credentials',
                                                            'client_id' : authItems['appId'],
                                                            'client_secret' : authItems['appKey'],
                                                            'resource' : authItems['resource']
                                                        }
                                            });
        };

        //parse urls
        this.getLocation = function(location){
        	var temp = document.createElement("a");
        	temp.href = location;
        	return temp.pathname;
        };

        //method to create QiTypes
        this.createType = function(wave){
                                return restCall({
                                                url : this.url+this.typesBase,
                                                method: 'POST',
                                                headers : this.getHeaders(),
                                                body : JSON.stringify(wave).toString()
                                            });
        };

        //method to get all the Qi types under a tenant Qi Service
        this.getTypes = function(){
                                return restCall({
                                                url : this.url+this.typesBase,
                                                method: 'GET',
                                                headers : this.getHeaders()
                                            });
        };

        //delete a type
        this.deleteType = function(typeId){
                                return restCall({
                                            url : this.url+this.typesBase+"/"+typeId,
                                            method : 'DELETE',
                                            headers : this.getHeaders()
                                        });
        };

        //create a stream under the Qi Service
        this.createStream = function(qiStream){
                                return restCall({
                                            url : this.url+this.streamsBase,
                                            method : 'POST',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(qiStream).toString()
                                        });
        };

        //get all the streams under the tenant's Qi Service
        this.getStream = function(qiStream){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+qiStream.Id,
                                            method : 'GET',
                                            headers : this.getHeaders()
                                        });
        };

        //get all the streams under the tenant's Qi Service
        this.getStreams = function(){
                                return restCall({
                                            url : this.url+this.streamsBase,
                                            method : 'GET',
                                            headers : this.getHeaders()
                                        });
        };

        //update a stream
        this.updateStream = function(qiStream){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+qiStream.Id,
                                            method : 'PUT',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(qiStream).toString()
                                        });
        };

        //delete a stream
        this.deleteStream = function(streamId){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+streamId,
                                            method : 'DELETE',
                                            headers : this.getHeaders()
                                        });
        };

        //create behavior
        this.createBehavior = function(behavior){
                                return restCall({
                                            url : this.url+this.behaviorBase,
                                            method : 'POST',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(behavior)
                                        });
        };

        //insert an event into a stream
        this.insertEvent = function(qiStream, evt){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+
                                                    qiStream.Id+this.insertSingle,
                                            method : 'POST',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(evt)
                                        });
        };

        //insert an array of events
        this.insertEvents = function(qiStream, events){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+
                                                    qiStream.Id+this.insertMultiple,
                                            method : 'POST',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(events)
                                        });
        };

        //update an event
        this.updateEvent = function(qiStream, evt){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+
                                                    qiStream.Id+this.updateSingle,
                                            method : 'PUT',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(evt)
                                        });
        };

        //update an array of events
        this.updateEvents = function(qiStream, events){
                                return restCall({
                                            url : this.url+this.streamsBase+"/"+
                                                    qiStream.Id+this.updateMultiple,
                                            method : 'PUT',
                                            headers : this.getHeaders(),
                                            body : JSON.stringify(events)
                                        });
        };

        //delete an event
        this.deleteEvent = function(qiStream, index){
                                return restCall({
                                            url : this.url+this.streamsBase+this.removeSingleTemplate.format([qiStream.Id, index]),
                                            method : 'DELETE',
                                            headers : this.getHeaders()
                                        });
        };

        //delete a window of events
        this.deleteWindowEvents = function(qiStream, start, end){
                                return restCall({
                                            url : this.url+this.streamsBase+this.removeMultipleTemplate.format([qiStream.Id, start, end]),
                                            method : 'DELETE',
                                            headers : this.getHeaders()
                                        });
        };

        //retrieve a window of events
        this.getWindowValues = function(qiStream, start, end){
                                return restCall({
                                            url : this.url+this.streamsBase+this.getTemplate.format([qiStream.Id,start,end]),
                                            method : 'GET',
                                            headers : this.getHeaders()
                                        });
        };

        //retrieve a range of value based on boundary type
        this.getRangeValues = function(qiStream, start, skip, count, reverse, boundaryType){
                                return restCall({
                                            url : this.url+this.streamsBase+this.getRangeTemplate.format([qiStream.Id,start,skip,count,reverse,boundaryType]),
                                            method : 'GET',
                                            headers : this.getHeaders()
                                        });
        };

        this.getHeaders = function(){
                                return {
                                            "Authorization" : "bearer "+ this.token,
                                            "Content-type": "application/json", 
                                            "Accept": "text/plain"
                                        }
        };
    }
};